### DISCUSSION TITLE
Value objects as query parameters?

### DISCUSSION THREAD

**Discussion by roengstrom** [reactiveui/refit#1633] [General]
Hello!

I'm trying to use Refit in my project where I want to be able to supply a value object as a query parameter. However it doesn't seem to work - it just omits the parameter. Is that expected behaviour or is it a bug?

Here is my value object:
```
public readonly record struct CustomerId
{
    private readonly string _customerId;
    
    public CustomerId(string customerId)
    {
        _customerId = customerId ?? throw new NullReferenceException("CustomerId cannot be null");
        // (More validation is done here, but I've omitted here since it's irrelevant to the issue)
    }
    
    public override string ToString()
    {
        return _customerId;
    }
}
```

And here is my refit interface
```
public interface IMyApi
{
    [Get("/Items")]
    Task<ICollection<Item>> GetItemsForCustomer(
        [Query] CustomerId customerId
    );
}
```

**Comment by glennawatson** [comment]
This is expected behaviour, not a bug. Your CustomerId is a record struct with no public properties (just a private _customerId field). For a query parameter, Refit decides between two paths:

- If the type is "simple" (string, bool, char, Uri, or anything implementing IFormattable), it is rendered directly via its string form.
- Otherwise Refit treats it as an object and expands its public readable properties into the query string.

CustomerId hits the second path, finds no public properties, so nothing is emitted - hence the missing parameter. Overriding ToString does not change which path is taken.

Fix: make the value object IFormattable so Refit renders it directly:

```csharp
public readonly record struct CustomerId : IFormattable
{
    private readonly string _customerId;
    public CustomerId(string customerId) =>
        _customerId = customerId ?? throw new ArgumentNullException(nameof(customerId));

    public override string ToString() => _customerId;
    public string ToString(string? format, IFormatProvider? formatProvider) => _customerId;
}
```

Alternatives:
- Add [Query(Format = "...")] handling via a custom IUrlParameterFormatter if you need centralized conversion for many value objects.
- Expose a public property (e.g. Value) if you actually want it expanded as customerId.Value=...
- Pass customerId.ToString() at the call site.

The IFormattable approach is the cleanest and matches how Refit already handles primitive query values.
